### PR TITLE
Make pvc storage class annotation immutable after create

### LINKED ISSUE #34516: PersistentVolumeClaim storage class is mutable
I am investigating the ability to quota `requests.storage` by `StorageClass`.

It appears that you can mutate `volume.beta.kubernetes.io/storage-class` post creation.

I think this field should work like 'requests.storage` and be immutable post creation.

### PR DESCRIPTION
**Pull request by screeley44**
Fixes #34516

Added immutable check in validation.go

```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# persistentvolumeclaims "gce-claim-storageclass" was not valid:
# * metadata.annotations.volume.beta.kubernetes.io/storage-class: Invalid value: "slow2": field is immutable
#

```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36251)

### CODE REVIEW

**Comment by screeley44** [comment]
@eparis @erinboyd @derekwaynecarr - how does this look?

**Comment by smarterclayton** [comment]
Add a test case.

**Comment by screeley44** [comment]
@smarterclayton - test case added

**Comment by screeley44** [comment]
@smarterclayton - what is that verify-bazel.sh and update-bazel.sh ?  I ran that and it modified a ton of files, am I supposed to commit each of those to get past that test?

**Comment by smarterclayton** [comment]
Yes, you must commit the bazel files.  Those are build system metadata files that go with each package.

**Comment by screeley44** [comment]
@k8s-bot test this issue: #31776

**Comment by screeley44** [comment]
@eparis @smarterclayton - PTAL

**Comment by eparis** [comment]
@kubernetes/sig-storage

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -351,6 +352,15 @@ func ValidateImmutableField(newVal, oldVal interface{}, fldPath *field.Path) fie
 	return allErrs
 }
 
+func ValidateImmutableAnnotation(newVal, oldVal interface{}, annotation string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !api.Semantic.DeepEqual(oldVal, newVal) {
```
Comment (jsafrane) [inline pkg/api/validation/validation.go]: annotation values are strings, no need for generic `interface{}` and `DeepEqual` here. Use `string` and simple `==`.

File: pkg/api/validation/validation_test.go [hunk 2]
```diff
@@ -874,6 +911,11 @@ func TestValidatePersistentVolumeClaimUpdate(t *testing.T) {
 			oldClaim:          validUpdateClaim,
 			newClaim:          invalidUpdateClaimAccessModes,
 		},
+		"invalid-update-change-storage-class-annotation-after-creation": {
```
Comment (jsafrane) [inline pkg/api/validation/validation_test.go]: Please add a test that updates another (=allowed) annotation - such update should be valid.

**Review by jsafrane — CHANGES_REQUESTED** [review]

File: pkg/api/validation/validation_test.go (hunk 2 above)
Comment (screeley44) [inline pkg/api/validation/validation_test.go]: @jsafrane 

I updated the tests - added a valid change to annotation and a test to make sure an annotation could be added

```
+++ [1116 12:03:37] Running tests without code coverage
ok  	k8s.io/kubernetes/pkg/api/validation	0.341s
```

**Review by jsafrane — APPROVED** [review]

**Comment by jsafrane** [comment]
lgtm. @smarterclayton PTAL, I'm going to merge it if you don't have any remarks.

**Comment by screeley44** [comment]
@jsafrane @smarterclayton - can this be merged?

**Comment by smarterclayton** [comment]
Please squash - you've got a merge commit in there and we don't want merge commits in the normal tree.  Removing lgtm

**Comment by jsafrane** [comment]
@k8s-bot verify test this

**Comment by screeley44** [comment]
@smarterclayton - we good now?

**Comment by smarterclayton** [comment]
Yes

**Comment by screeley44** [comment]
@jsafrane - can we merge this now?

**Comment by jsafrane** [comment]
@screeley44, it's waiting in merge queue. Click on "Submit Queue" in the window with test results to see how it stands. It's a long queue!

**Pull request merged**